src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,494 @@
+            if (!isImmutable(item)) {
+                final String message = String
+                    .format(Locale.ROOT, "Field <%s> should %s in %s",
+                            item.getFullName(), getDescription(),
+                            item.getSourceCodeLocation().toString());
+                events.add(SimpleConditionEvent.violated(item, message));
+            }
+        }
+    }
+
+    /**
+     * DescribedPredicate defining condition for a field to be a module property.
+     */
+    private static final class ModulePropertyPredicate extends DescribedPredicate<JavaField> {
+
+        private ModulePropertyPredicate() {
+            super("module properties");
+        }
+
+        /**
+         * Whether a field is a module property or not.
+         *
+         * @param javaField field to check
+         * @return {@code true} if field is a module property
+         */
+        private static boolean isModuleProperty(JavaField javaField) {
+            boolean result = false;
+            final JavaClass containingClass = javaField.getOwner();
+            final ModuleDetails moduleDetails = MODULE_DETAILS_MAP.get(
+                containingClass.getFullName());
+            if (moduleDetails != null) {
+                final List<ModulePropertyDetails> properties = moduleDetails.getProperties();
+                result = properties.stream()
+                    .map(ModulePropertyDetails::getName)
+                    .anyMatch(moduleName -> moduleName.equals(javaField.getName()));
+            }
+            return result;
+        }
+
+        @Override
+        public boolean apply(JavaField input) {
+            return isModuleProperty(input);
+        }
+    }
+
+    /**
+     * DescribedPredicate defining condition for a class to have immutable fields.
+     */
+    private static final class ImmutableFieldsPredicate extends DescribedPredicate<JavaClass> {
+        private ImmutableFieldsPredicate() {
+            super("immutable fields");
+        }
+
+        @Override
+        public boolean apply(JavaClass input) {
+            final Set<JavaField> fields = input.getFields();
+            return fields.stream()
+                .filter(javaField -> {
+                    return !ModulePropertyPredicate.isModuleProperty(javaField)
+                        && !SUPPRESSED_FIELDS_IN_MODULES.contains(javaField.getFullName());
+                })
+                .allMatch(javaField -> {
+                    final Set<JavaModifier> javaFieldModifiers = javaField.getModifiers();
+                    boolean result = javaFieldModifiers.contains(JavaModifier.FINAL)
+                        && ImmutableFieldArchCondition.isImmutable(javaField);
+                    // 'MSG_'s in checks are public as they are imported in test files.
+                    if (!javaField.getName().contains("MSG_")) {
+                        result = result && javaFieldModifiers.contains(JavaModifier.PRIVATE);
```

Comment:
Connected to an earlier comment, I don't think private and immutable should be directly connected.

Reply:
Fields should be private see https://stackoverflow.com/q/28410386/14425335

Reply:
https://stackoverflow.com/q/28410386/14425335
> That is ONE reason why it is a good idea to keep the fields private. (**Obviously, this only applies to field whose types are mutable reference types.**)

Fields must be immutable for Stateless, regardless of modifiers.

Reply:
I removed the thing about checking for public, private, should be fine now.